python/ngen_cal/src/ngen/cal/__main__.py

```diff
@@ -46,21 +77,31 @@ def main(general: General, model_conf: "Mapping[str, Any]"):
     # print("Starting Best score: {}".format(meta.best_score))
     print("Starting calibration loop")
 
+    # call `ngen_cal_start` plugin hook functions
+    plugin_manager.hook.ngen_cal_start()
+
+    try:
+        #NOTE this assumes we calibrate each catchment independently, it may be possible to design an "aggregate" calibration
+        #that works in a more sophisticated manner.
+        if agent.model.strategy == 'explicit': #FIXME this needs a refactor...should be able to use a calibration_set with explicit loading
+            for catchment in agent.model.adjustables:
+                dds(start_iteration, general.iterations, catchment, agent)
 
-    #NOTE this assumes we calibrate each catchment independently, it may be possible to design an "aggregate" calibration
-    #that works in a more sophisticated manner.
-    if agent.model.strategy == 'explicit': #FIXME this needs a refactor...should be able to use a calibration_set with explicit loading
-        for catchment in agent.model.adjustables:
-            dds(start_iteration, general.iterations, catchment, agent)
+        elif agent.model.strategy == 'independent':
+            #for catchment_set in agent.model.adjustables:
+            func(start_iteration, general.iterations, agent)
 
-    elif agent.model.strategy == 'independent':
-        #for catchment_set in agent.model.adjustables:
-        func(start_iteration, general.iterations, agent)
+        elif agent.model.strategy == 'uniform':
+            #for catchment_set in agent.model.adjustables:
+            #    func(start_iteration, general.iterations, catchment_set, agent)
+            func(start_iteration, general.iterations, agent)
+    # call `ngen_cal_finish` plugin hook functions
+    except Exception as e:
+        plugin_manager.hook.ngen_cal_finish(exception=e)
+        raise e
```

Comment:
Hmm...what happens here if the plugin raises an exception?  Do we need to nest a try and report both the plugin exception AND the ngen-cal exception?

Reply:
Im fine with the plugin raising an exception. IMO that is not our code so it is their responsibility to handle exceptions.